// yarushawombater-lb3/webpack.config.mjs
import path from "path";
import { fileURLToPath } from "url";
import HtmlWebpackPlugin from 'html-webpack-plugin';
import { CleanWebpackPlugin } from 'clean-webpack-plugin';
import fs from "fs";

const __filename = fileURLToPath(import.meta.url);
const __dirnamePath = path.dirname(__filename);

// Загрузка данных пользователей и новостей
const users = JSON.parse(fs.readFileSync(path.join(__dirnamePath, "data/users.json"), "utf8"));
const friends = users.filter(u => users[0].friends.includes(u.id));
const friendIds = friends.map(f => f.id);
const news = JSON.parse(fs.readFileSync(path.join(__dirnamePath, "data/news.json"), "utf8"))
    .filter(news => friendIds.includes(news.userId));

const newsData = news.map(news => {
    const friend = users.find(u => u.id === news.userId);
    return {
        ...news,
        fullName: friend ? friend.fullName : 'Неизвестный'
    };
});

export default {
    entry: {
        main: './public/js/main.js',
        mainTemplate: './views/layouts/main.pug',
        usersTemplate: './views/admin/users.pug',
        editUserTemplate: './views/admin/editUser.pug',
        friendsTemplate: './views/admin/friends.pug',
        newsTemplate: './views/admin/news.pug',
        errorTemplate: './views/error.pug',
    },
    module: {
        rules: [
            {
                test: /\.js$/,
                exclude: /node_modules/,
                use: "babel-loader"
            },
            {
                test: /\.pug$/,
                loader: "pug-loader",
                options: {
                    pretty: true
                }
            },
            {
                test: /\.scss$/,
                use: [
                    'style-loader', // Вставка стилей в DOM
                    'css-loader',   // Преобразование CSS в CommonJS
                    'sass-loader'   // Компиляция SCSS в CSS
                ]
            }
        ]
    },
    plugins: [
        new CleanWebpackPlugin(), // Очистка папки dist/webpack перед сборкой
        new HtmlWebpackPlugin({
            filename: 'mainTemplate.html',
            template: './views/layouts/main.pug',
            inject: 'body',
            chunks: ['main', 'mainTemplate'],
            templateParameters: {users: users}
        }),
        new HtmlWebpackPlugin({
            filename: 'usersTemplate.html',
            template: './views/admin/users.pug',
            inject: 'body',
            chunks: ['main', 'usersTemplate'],
            templateParameters: {users: users}
        }),
        new HtmlWebpackPlugin({
            filename: 'editUserTemplate.html',
            template: './views/admin/editUser.pug',
            inject: 'body',
            chunks: ['main', 'editUserTemplate'],
            templateParameters: {user: users[0]}
        }),
        new HtmlWebpackPlugin({
            filename: 'friendsTemplate.html',
            template: './views/admin/friends.pug',
            inject: 'body',
            chunks: ['main', 'friendsTemplate'],
            templateParameters: {user: users[0], friends: friends}
        }),
        new HtmlWebpackPlugin({
            filename: 'newsTemplate.html',
            template: './views/admin/news.pug',
            inject: 'body',
            chunks: ['main', 'newsTemplate'],
            templateParameters: {user: users[0], newsData: newsData}
        }),
        new HtmlWebpackPlugin({
            filename: 'errorTemplate.html',
            template: './views/error.pug',
            inject: 'body',
            chunks: ['main', 'errorTemplate']
        }),
    ],
    resolve: {
        extensions: ['.js', '.scss', '.pug']
    },
    output: {
        path: path.resolve(__dirnamePath, "dist/webpack/"),
        filename: '[name].bundle.js',
    },
    mode: 'development'
};
